### DISCUSSION TITLE
Does Pandoc follow symlinks when looking for Lua filters?

### DISCUSSION THREAD

**Discussion by johnmapeson** [jgm/pandoc#10432] [Q&A] [answered]
Pandoc 3.4, macOS 14.7.1

I have a Lua filter in my Pandoc data directory, which is `/Users/jsx/.local/share/pandoc/filters`, and it works. Then I moved that filter to my `$HOME/scripts` directory, which is already added to `$PATH`, and created a symlink from Pandoc data directory to there:

```
ln -s $HOME/scripts /Users/jsx/.local/share/pandoc/filters
```

But It seems that in such a case Pandoc won't see a filter:

```
cd ~/test
pandoc -o output.md -L my-filter.lua input.htm
```

```
Error running filter my-filter.lua:
cannot open my-filter.lua: No such file or directory
```

Am I correct? (Or maybe I simly don't use symlinks correctly.)

**Comment by jgm** [comment]
This should work.  Doing something similar worked fine for me in a test I just did.
If you do `ls -l $HOME/.local/share/pandoc`, does `filters` show up as a symlink?
Do you see the right things with `ls $HOME/.local/share/pandoc/filters`?

**Comment by johnmapeson** [comment]
### Here is a detailed walkthrough of what I tried:

1. Make sure the folder `/Users/jsx/scripts` is added to `$PATH`.

  `echo $PATH`:

  ```
  <skipped for brevity>:/Users/jsx/scripts:<skipped for brevity>
  ```

2. Make sure the filter is there (that is, in `/Users/jsx/scripts`).

3. `cd` to another directory, and run the `pandoc` command there.

   ```
   cd ~/test
   pandoc -o output.md -L filter.lua input.htm
   ```

   ```
   Error running filter filter.lua:
   cannot open filter.lua: No such file or directory
   ```

### And here is an attempt to figure out what is going on:

In Finder, the contents of the `/Users/jsx/.local/share/pandoc/filters` folder looks like this:

<img width="308" alt="image" src="https://github.com/user-attachments/assets/d914edff-174d-49ea-8f87-b3a8d0509268">

The icon is a symlink to `/Users/jsx/scripts`.

The output of `ls $HOME/.local/share/pandoc/filters` is just

```
scripts
```

The output of `ls -l $HOME/.local/share/pandoc` is

```
total 0
drwxr-xr-x  4 jsx  staff  128 Dec  2 21:43 filters
```

The output of `ls -l $HOME/.local/share/pandoc/filters` is

```
total 0
lrwxr-xr-x  1 jsx  staff  19 Dec  2 21:31 scripts -> /Users/jsx/scripts
```

**Reply by jgm** [reply, marked as answer]
OK, looks like you created a symlink *within* `filters` to your scripts directory.  But pandoc doesn't look in subdirectories of the `filters` directory.  You could try:
```
rm $HOME/.local/share/pandoc/filters/scripts
rmdir $HOME/.local/share/pandoc/filters
ln -s $HOME/scripts $HOME/.local/share/pandoc/filters
```
That makes the `filters` directory point directry to your `scripts` directory.

Alternatively it may work to do what you did but use `-L scripts/filter.lua` (untested).

**Comment by johnmapeson** [comment]
John, I know you have more important things to do, but I've been trying to figure this out on my own for about two weeks - and haven't had any success yet. Two weeks ago, you suggested me:

```
rm $HOME/.local/share/pandoc/filters/scripts
rmdir $HOME/.local/share/pandoc/filters
ln -s $HOME/scripts $HOME/.local/share/pandoc/filters
```

And it worked fine. But then, at some moment, the shell started to complain about missing filters again.

The current output of `ls -l $HOME/.local/share/pandoc/filters`:

```
/Users/myname/.local/share/pandoc/filters@ -> /Users/myname/usr
```

(Here, `usr` is a custom directory I use for my personal command-line related stuff. It is added it to `$PATH`, of course.)

Inside `usr`, I have several pandoc filters, e.g. `foo.lua` and `bar.lua`.

Then I do

```
cd ~/test
pandoc -o output.md -L my-filter.lua input.htm
```

I receive that "cannot open foo.lua: No such file or directory" error again.

What might be going on here? 🤔

**Reply by jgm** [reply]
If you do `ls ~/.local/share/pandoc/filters`, do you see `foo.lua`?

**Reply by johnmapeson** [reply]
Yes.
